From 3557cd57906915eb9c990b5f386e25c395592643 Mon Sep 17 00:00:00 2001
From: Paul Eggert <eggert@cs.ucla.edu>
Date: Wed, 20 Apr 2016 00:43:09 +0000
Subject: gzip: fix bug with -l output to pipes

Problem reported by Christian Franke via Eric Blake in:
http://bugs.gnu.org/23314
* NEWS: Mention this.
* gzip.c (main): Do not close stdout twice when given -l.
Instead, -l now just fflushes stdout, so that fdatasync
can synchronize it if --synchronize is also specified.
* tests/list: New test case.
* tests/Makefile.am (TESTS): Add it.
---
diff --git a/gzip.c b/gzip.c
index d66530a..3b8de4d 100644
--- a/gzip.c
+++ b/gzip.c
@@ -664,14 +664,16 @@ int main (int argc, char **argv)
     } else {  /* Standard input */
         treat_stdin();
     }
-    if (list && !quiet && file_count > 1) {
-        do_list(-1, -1); /* print totals */
-    }
-    if (to_stdout
-        && ((synchronous
-             && (fdatasync (STDOUT_FILENO) != 0 && errno != EINVAL))
-            || close (STDOUT_FILENO) != 0)
-        && errno != EBADF)
+    if (list)
+      {
+        /* Output any totals, and check for output errors.  */
+        if (!quiet && 1 < file_count)
+          do_list (-1, -1);
+        if (fflush (stdout) != 0)
+          write_error ();
+      }
+    if (to_stdout && synchronous && fdatasync (STDOUT_FILENO) != 0
+        && errno != EINVAL && errno != EBADF)
       write_error ();
     do_exit(exit_code);
     return exit_code; /* just to avoid lint warning */
--
cgit v0.9.0.2
